x-pack/plugins/remote_clusters/public/application/sections/components/remote_cluster_form/remote_cluster_form.tsx

```diff
@@ -0,0 +1,630 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import React, { Component, Fragment } from 'react';
+import { merge } from 'lodash';
+import { i18n } from '@kbn/i18n';
+import { FormattedMessage } from '@kbn/i18n/react';
+import {
+  EuiButton,
+  EuiButtonEmpty,
+  EuiCallOut,
+  EuiDescribedFormGroup,
+  EuiFieldText,
+  EuiFlexGroup,
+  EuiFlexItem,
+  EuiForm,
+  EuiFormRow,
+  EuiLink,
+  EuiLoadingKibana,
+  EuiLoadingSpinner,
+  EuiOverlayMask,
+  EuiSpacer,
+  EuiSwitch,
+  EuiText,
+  EuiTitle,
+  EuiDelayRender,
+  EuiScreenReaderOnly,
+  htmlIdGenerator,
+  EuiSwitchEvent,
+} from '@elastic/eui';
+
+import { Cluster } from '../../../../../common/lib';
+import { SNIFF_MODE, PROXY_MODE } from '../../../../../common/constants';
+
+import { AppContext, Context } from '../../../app_context';
+
+import { skippingDisconnectedClustersUrl } from '../../../services/documentation';
+
+import { RequestFlyout } from './request_flyout';
+import { ConnectionMode } from './connection_mode';
+import {
+  ClusterErrors,
+  validateCluster,
+  convertCloudUrlToProxyConnection,
+  convertProxyConnectionToCloudUrl,
+  isCloudUrl,
+} from './validators';
+
+const defaultClusterValues: Cluster = {
+  name: '',
+  seeds: [],
+  skipUnavailable: false,
+  nodeConnections: 3,
+  proxyAddress: '',
+  proxySocketConnections: 18,
+  serverName: '',
+};
+
+const ERROR_TITLE_ID = 'removeClustersErrorTitle';
+const ERROR_LIST_ID = 'removeClustersErrorList';
+
+interface Props {
+  save: (cluster: Cluster) => void;
+  cancel?: () => void;
+  isSaving?: boolean;
+  saveError?: any;
+  cluster?: Cluster;
+}
+
+export type FormFields = Cluster & { isCloudUrl: boolean; cloudUrl: string };
+
+interface State {
+  fields: FormFields;
+  fieldsErrors: ClusterErrors;
+  areErrorsVisible: boolean;
+  isRequestVisible: boolean;
+}
+
+export class RemoteClusterForm extends Component<Props, State> {
+  static defaultProps = {
+    fields: merge({}, defaultClusterValues),
+  };
+
+  static contextType = AppContext;
+  private readonly generateId: (idSuffix?: string) => string;
+
+  constructor(props: Props, context: Context) {
+    super(props, context);
+
+    const { cluster } = props;
+    const { isCloudEnabled } = context;
+
+    // Connection mode should default to "proxy" in cloud
+    const defaultMode = isCloudEnabled ? PROXY_MODE : SNIFF_MODE;
+    const isCloudUrlEnabled: boolean = isCloudUrl(cluster) && isCloudEnabled;
+    const fieldsState: FormFields = merge(
+      {},
+      {
+        ...defaultClusterValues,
+        mode: defaultMode,
+        isCloudUrl: isCloudUrlEnabled,
```

Comment:
Thanks for bringing this up, @alisonelizabeth! We still need to track it in the form component if the cloud url is being used. Otherwise, the form component won't be able to validate the cluster input correctly. But I refactored the `connection_mode` component into smaller parts to make it a little clearer, I hope. Would you mind having another look, please?

Reply:
👍 Thanks for addressing this. The latest implementation is much easier to follow.